owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringState.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package com.facebook.openwifi.rrm.optimizers.clientsteering;
+
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/** Singleton class to manage global client steering state */
+public final class ClientSteeringState {
+
+	/** Singleton instance */
+	private static final ClientSteeringState singleton =
+		new ClientSteeringState();
+
+	/** Private default constructor (this is a singleton class) */
+	private ClientSteeringState() {};
+
+	/** Get the singleton instance */
+	public static ClientSteeringState getInstance() {
+		return singleton;
+	}
+
+	/**
+	 * Map from AP serial number to radio bssid to time (JVM monotonic time in
+	 * ms) of the latest attempted client steering action. The {@code Long}
+	 * values are never null.
+	 */
+	private final ConcurrentMap<String, Map<String, Long>> apRadioLastAttempt =
+		new ConcurrentHashMap<>();
+
+	/** Reset the state (e.g., for testing) */
+	public final void reset() {
+		apRadioLastAttempt.clear();
+	}
+
+	/**
+	 * Register the time of the latest client steering attempt by the given AP
+	 * and radio.
+	 *
+	 * @param apSerialNumber non-null AP serial number
+	 * @param bssid non-null radio bssid
+	 * @param currentTimeNs current JVM monotonic time (ns)
+	 */
+	final void registerClientSteeringAttempt(
```

Comment:
- Why are all of these `final`?
- Missing `public`?
- You need to think about how multiple concurrent `ClientSteeringOptimizer` instances might be calling these methods - hopefully not on the same AP simultaneously, but you may either need to handle that case or explicitly disallow it.

Reply:
They don't have to be final, wasn't sure about that. I didn't want it public as I figured nothing outside this package would need to be considered with this (I was going to remove `public` from the other methods too, i forgot).

Regarding concurrency, I'm not sure the best way to resolve this. Does it make sense to just make `ClientSteeringOptimizer#computeApClientActionMap` `synchronized` and make sure (and leave a comment) not to access `ClientSteeringState#registerClientSteeringAttempt` anywhere except in implementations of that method?

Reply:
IMO keep `public` and drop the `final`. Don't overcomplicate this unless there is a need.

For the methods, I am more concerned about interleaving "getLatest", "checkBackoff", "register" methods from the caller. Think about "transactional" operations, which might mean changing the methods or doing locking on keys (AP, client, etc.).